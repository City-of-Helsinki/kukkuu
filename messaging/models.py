from random import random

from django.db import models
from django.utils.timezone import now
from django.utils.translation import gettext_lazy as _
from parler.models import TranslatedFields
from projects.models import Project

from common.models import TimestampedModel, TranslatableModel, TranslatableQuerySet
from events.models import Event, Occurrence
from users.models import Guardian


class MessageQuerySet(TranslatableQuerySet):
    def user_can_view(self, user):
        return self.filter(project__users=user)


class Message(TimestampedModel, TranslatableModel):
    ALL = "all"
    INVITED = "invited"
    ENROLLED = "enrolled"
    ATTENDED = "attended"
    SUBSCRIBED_TO_FREE_SPOT_NOTIFICATION = "subscribed_to_free_spot_notification"
    RECIPIENT_SELECTION_CHOICES = (
        (ALL, _("All")),
        (INVITED, _("Invited")),
        (ENROLLED, _("Enrolled")),
        (ATTENDED, _("Attended")),
        (
            SUBSCRIBED_TO_FREE_SPOT_NOTIFICATION,
            _("Subscribed to free spot notification"),
        ),
    )

    project = models.ForeignKey(
        Project,
        verbose_name=_("project"),
        related_name="messages",
        on_delete=models.CASCADE,
    )
    translations = TranslatedFields(
        subject=models.CharField(verbose_name=_("subject"), max_length=255),
        body_text=models.TextField(verbose_name=_("body plain text")),
    )
    sent_at = models.DateTimeField(verbose_name=_("sent at"), blank=True, null=True)
    recipient_count = models.PositiveIntegerField(
        verbose_name=_("recipient count"), blank=True, null=True
    )
    recipient_selection = models.CharField(
        max_length=64,
        verbose_name=_("recipient selection"),
        choices=RECIPIENT_SELECTION_CHOICES,
    )
    event = models.ForeignKey(
        Event,
        verbose_name=_("event"),
        related_name="messages",
        blank=True,
        null=True,
        on_delete=models.CASCADE,
    )
    occurrences = models.ManyToManyField(
        Occurrence, verbose_name=_("occurrences"), related_name="messages", blank=True,
    )

    objects = MessageQuerySet.as_manager()

    class Meta:
        ordering = ("id",)
        verbose_name = _("message")
        verbose_name_plural = _("messages")

    def __str__(self):
        return f"({self.pk}) {self.subject} ({self.sent_at or 'not sent'})"

    def send(self):
        self.recipient_count = (
            Guardian.objects.filter(children__project=self.project).count()
            if self.recipient_selection == Message.ALL
            else random.randint(
                0, 50
            )  # TODO temporary thingy before actual send implementation
        )
        self.sent_at = now()
        self.save(update_fields=("recipient_count", "sent_at"))

    def can_user_administer(self, user):
        return user.projects.filter(pk=self.project.pk).exists()
