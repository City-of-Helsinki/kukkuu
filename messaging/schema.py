import logging
from copy import deepcopy

import graphene
from django.apps import apps
from django.db import transaction
from graphene import relay
from graphene_django import DjangoObjectType
from graphene_django.filter import DjangoFilterConnectionField

from common.schema import LanguageEnum
from common.utils import (
    get_obj_if_user_can_administer,
    project_user_required,
    update_object_with_translations,
)
from events.models import Event, Occurrence
from kukkuu.exceptions import DataValidationError, MessageAlreadySentError
from projects.models import Project

from .models import AlreadySentError, Message

logger = logging.getLogger(__name__)

MessageTranslation = apps.get_model("messaging", "MessageTranslation")


class MessageTranslationType(DjangoObjectType):
    class Meta:
        model = MessageTranslation
        exclude = ("id", "master")


class RecipientSelectionEnum(graphene.Enum):
    ALL = "all"
    INVITED = "invited"
    ENROLLED = "enrolled"
    ATTENDED = "attended"
    SUBSCRIBED_TO_FREE_SPOT_NOTIFICATION = "subscribed_to_free_spot_notification"


class MessageNode(DjangoObjectType):
    subject = graphene.String()
    body_text = graphene.String()
    recipient_selection = graphene.Field(RecipientSelectionEnum)

    class Meta:
        model = Message
        interfaces = (relay.Node,)
        fields = (
            "id",
            "project",
            "created_at",
            "updated_at",
            "subject",
            "body_text",
            "recipient_selection",
            "event",
            "occurrences",
            "sent_at",
            "recipient_count",
            "translations",
        )
        filter_fields = ("project_id",)

    @classmethod
    @project_user_required
    def get_queryset(cls, queryset, info):
        return queryset.user_can_view(info.context.user).prefetch_related(
            "translations"
        )

    @classmethod
    @project_user_required
    def get_node(cls, info, id):
        try:
            return (
                cls._meta.model.objects.user_can_view(info.context.user)
                .prefetch_related("translations")
                .get(id=id)
            )
        except cls._meta.model.DoesNotExist:
            return None

    def resolve_recipient_count(self, info, **kwargs):
        return self.get_recipient_count()


class MessageTranslationsInput(graphene.InputObjectType):
    language_code = LanguageEnum(required=True)
    subject = graphene.String()
    body_text = graphene.String()


class AddMessageMutation(graphene.relay.ClientIDMutation):
    class Input:
        translations = graphene.List(MessageTranslationsInput)
        project_id = graphene.GlobalID()
        recipient_selection = RecipientSelectionEnum(required=True)
        event_id = graphene.ID()
        occurrence_ids = graphene.List(graphene.NonNull(graphene.ID))

    message = graphene.Field(MessageNode)

    @classmethod
    @project_user_required
    @transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        data = deepcopy(kwargs)

        data["project"] = get_obj_if_user_can_administer(
            info, data.pop("project_id"), Project
        )

        event_id = data.pop("event_id", None)
        if event_id:
            data["event"] = get_obj_if_user_can_administer(info, event_id, Event)
        else:
            data["event"] = None
        occurrences = [
            get_obj_if_user_can_administer(info, occurrence_id, Occurrence)
            for occurrence_id in data.pop("occurrence_ids", [])
        ]

        validate_recipient_selection_and_data(
            data["recipient_selection"], data["event"], occurrences
        )
        validate_event_and_occurrences(data["event"], occurrences)

        message = Message.objects.create_translatable_object(**data)
        if occurrences:
            message.occurrences.set(occurrences)

        logger.info(
            f"user {info.context.user.uuid} added message {message} with data {kwargs}"
        )

        return AddMessageMutation(message=message)


class UpdateMessageMutation(graphene.relay.ClientIDMutation):
    class Input:
        id = graphene.GlobalID()
        translations = graphene.List(MessageTranslationsInput)
        project_id = graphene.ID()
        recipient_selection = RecipientSelectionEnum()
        event_id = graphene.ID()
        occurrence_ids = graphene.List(graphene.NonNull(graphene.ID))

    message = graphene.Field(MessageNode)

    @classmethod
    @project_user_required
    @transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        data = deepcopy(kwargs)
        message = get_obj_if_user_can_administer(info, data.pop("id"), Message)

        if message.sent_at:
            raise MessageAlreadySentError(
                "Cannot update because the message has already been sent."
            )

        if "project_id" in data:
            data["project_id"] = get_obj_if_user_can_administer(
                info, data.pop("project_id"), Project
            )

        if "event_id" in data:
            event_id = data.pop("event_id")
            data["event"] = (
                get_obj_if_user_can_administer(info, event_id, Event)
                if event_id
                else None
            )
        if "occurrence_ids" in data:
            occurrences = [
                get_obj_if_user_can_administer(info, occurrence_id, Occurrence)
                for occurrence_id in data.pop("occurrence_ids", [])
            ]
        else:
            occurrences = None

        if recipient_selection := data.get("recipient_selection"):
            validate_recipient_selection_and_data(
                recipient_selection,
                data["event"] if "event" in data else message.event,
                occurrences,
            )
        validate_event_and_occurrences(
            data["event"] if "event" in data else message.event,
            occurrences if occurrences is not None else message.occurrences.all(),
        )

        update_object_with_translations(message, data)

        if occurrences is not None:
            message.occurrences.set(occurrences)

        logger.info(
            f"user {info.context.user.uuid} updated message {message} with data {kwargs}"  # noqa: E501
        )

        return UpdateMessageMutation(message=message)


def validate_recipient_selection_and_data(recipient_selection, event, occurrences):
    if recipient_selection == RecipientSelectionEnum.INVITED and (event or occurrences):
        raise DataValidationError(
            "Selecting an event or occurrences are not supported when "
            "recipient_selection is INVITED."
        )


def validate_event_and_occurrences(event, occurrences):
    if not occurrences:
        return

    if not event:
        raise DataValidationError("Event is needed when there are occurrences.")

    if any(occurrence.event != event for occurrence in occurrences):
        raise DataValidationError("All of the occurrences do not belong to the event.")


class SendMessageMutation(graphene.relay.ClientIDMutation):
    class Input:
        id = graphene.GlobalID()

    message = graphene.Field(MessageNode)

    @classmethod
    @project_user_required
    def mutate_and_get_payload(cls, root, info, **kwargs):
        message = get_obj_if_user_can_administer(info, kwargs["id"], Message)

        try:
            message.send()
        except AlreadySentError:
            raise MessageAlreadySentError(
                "Cannot send because the message has already been sent."
            )

        logger.info(f"user {info.context.user.uuid} sent message {message}")

        return SendMessageMutation(message=message)


class DeleteMessageMutation(graphene.relay.ClientIDMutation):
    class Input:
        id = graphene.GlobalID()

    @classmethod
    @project_user_required
    def mutate_and_get_payload(cls, root, info, **kwargs):
        message = get_obj_if_user_can_administer(info, kwargs["id"], Message)
        if message.sent_at:
            raise MessageAlreadySentError(
                "Cannot delete because the message has already been sent."
            )

        log_text = f"user {info.context.user.uuid} deleted message {message}"
        message.delete()

        logger.info(log_text)

        return DeleteMessageMutation()


class Query:
    message = relay.Node.Field(MessageNode)
    messages = DjangoFilterConnectionField(MessageNode)


class Mutation:
    add_message = AddMessageMutation.Field()
    update_message = UpdateMessageMutation.Field()
    send_message = SendMessageMutation.Field()
    delete_message = DeleteMessageMutation.Field()
