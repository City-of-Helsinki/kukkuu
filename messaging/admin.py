from django.contrib import admin
from django.utils.translation import gettext_lazy as _
from parler.admin import TranslatableAdmin

from .models import AlreadySentError, Message


@admin.register(Message)
class MessageAdmin(TranslatableAdmin):
    list_display = (
        "subject",
        "project",
        "sent_at",
        "get_recipient_count",
        "created_at",
        "updated_at",
    )
    fields = (
        "project",
        "subject",
        "body_text",
        "recipient_selection",
        "event",
        "occurrences",
        "created_at",
        "updated_at",
        "get_recipient_count",
    )
    readonly_fields = ("get_recipient_count", "created_at", "updated_at")
    actions = ("send",)

    def get_recipient_count(self, obj):
        return str(obj.get_recipient_count()) + (
            str(_(" (if sent now)")) if not obj.sent_at else ""
        )

    get_recipient_count.short_description = _("recipient count")

    def send(self, request, queryset):
        sent_count = 0
        for obj in queryset:
            try:
                obj.send()
                sent_count += 1
            except AlreadySentError:
                pass
        self.message_user(request, _("%s successfully sent.") % sent_count)
