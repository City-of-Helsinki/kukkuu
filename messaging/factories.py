import factory
from projects.models import Project

from .models import Message


class MessageFactory(factory.django.DjangoModelFactory):
    subject = factory.Faker("text", max_nb_chars=20)
    body_text = factory.Faker("paragraph", nb_sentences=5)
    project = factory.LazyFunction(lambda: Project.objects.get_or_create(year=2020)[0])
    recipient_selection = Message.ALL

    class Meta:
        model = Message
