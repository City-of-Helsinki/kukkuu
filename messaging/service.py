from typing import TYPE_CHECKING

from django.conf import settings
from django.utils import timezone
from parler.utils.context import switch_language

from django_ilmoitin.utils import Message as MailerMessage
from django_ilmoitin.utils import send_all, send_mail
from kukkuu.notification_service import send_sms_notification
from messaging.exceptions import AlreadySentError
from users.models import Guardian

if TYPE_CHECKING:
    from messaging.models import Message


def send_message(message: "Message", *, force=False):
    from messaging.models import Message

    if message.sent_at and not force:
        raise AlreadySentError()

    guardians = message.get_recipient_guardians()

    message.sent_at = timezone.now()
    message.recipient_count = len(guardians)

    message.save(update_fields=("sent_at", "recipient_count"))

    for guardian in guardians:
        with switch_language(message, guardian.language):
            if message.protocol == Message.EMAIL:
                send_email_notification(guardian, message)
            elif message.protocol == Message.SMS:
                send_sms_notification([guardian.phone_number], message.body_text)

    if not getattr(settings, "ILMOITIN_QUEUE_NOTIFICATIONS", False):
        MailerMessage.objects.retry_deferred()
        send_all()


def send_email_notification(guardian: Guardian, message: "Message"):
    # hopefully this functionality that uses django-ilmoitin's internals
    # is only temporarily here, and will be removed when either
    # 1) we need to use a notification template and thus will use
    #    ilmoitin's regular send_notification(), or
    # 2) support for sending a mail without a notification template will
    #    be added to ilmoitin and we can use that
    if guardian.language in getattr(settings, "ILMOITIN_TRANSLATED_FROM_EMAIL", {}):
        from_email = settings.ILMOITIN_TRANSLATED_FROM_EMAIL[guardian.language]
    else:
        from_email = settings.DEFAULT_FROM_EMAIL

    send_mail(
        message.subject,
        message.body_text,
        guardian.email,
        from_email=from_email,
    )
