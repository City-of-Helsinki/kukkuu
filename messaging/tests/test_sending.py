from datetime import timedelta

import pytest
from django.core import mail
from django.utils import timezone
from django.utils.timezone import now
from freezegun import freeze_time

from children.factories import ChildWithGuardianFactory
from common.tests.utils import assert_mails_match_snapshot
from events.factories import (
    EnrolmentFactory,
    EventFactory,
    EventGroupFactory,
    OccurrenceFactory,
)
from messaging.factories import MessageFactory
from messaging.models import AlreadySentError, Message
from projects.factories import ProjectFactory
from subscriptions.factories import FreeSpotNotificationSubscriptionFactory


@pytest.mark.parametrize("guardian_language", ("fi", "en"))
def test_message_sending(snapshot, guardian_language, message):
    ChildWithGuardianFactory(
        relationship__guardian__email="johndoe@example.com",
        relationship__guardian__language=guardian_language,
    )

    message.send()

    assert_mails_match_snapshot(snapshot)
    assert message.recipient_count == 1
    assert message.sent_at


def test_cannot_send_message_again_without_force(snapshot, message):
    message.sent_at = now()
    message.save(update_fields=("sent_at",))
    ChildWithGuardianFactory(relationship__guardian__email="johndoe@example.com")

    with pytest.raises(AlreadySentError):
        message.send()

    assert len(mail.outbox) == 0

    message.send(force=True)

    assert_mails_match_snapshot(snapshot)
    assert message.recipient_count == 1
    assert message.sent_at


@freeze_time("2020-11-11")
@pytest.mark.parametrize(
    "recipient_selection",
    (
        Message.ALL,
        Message.INVITED,
        Message.ATTENDED,
        Message.ENROLLED,
        Message.SUBSCRIBED_TO_FREE_SPOT_NOTIFICATION,
    ),
)
@pytest.mark.parametrize(
    "event_selection",
    (
        "no_event_or_occurrence",
        "event",
        "occurrence_tomorrow_1",
        "occurrence_yesterday_1",
    ),
)
def test_message_sending_with_filters(snapshot, recipient_selection, event_selection):
    """Ridiculously complex filtering test

    There are three events:
        * event with 4 occurrences (2 in the past and 2 in the future):
            * tomorrow_1 with 1 enrolled child and 1 subscribed child
            * yesterday_1 with 1 attended child
            * tomorrow_2
            * yesterday_2
        * another event with 1 occurrence with 1 enrolled child and 1 subscribed child
        * event from another project with 1 enrolled child

    With those we test all the combinations of recipient_selection and
    event/occurrences selection.
    """
    yesterday = timezone.now() - timedelta(days=1)
    tomorrow = timezone.now() + timedelta(days=1)

    event = EventFactory(published_at=timezone.now())

    occurrence_tomorrow_1 = OccurrenceFactory(event=event, time=tomorrow)
    occurrence_yesterday_1 = OccurrenceFactory(event=event, time=yesterday)

    child_enrolled_tomorrow_1 = ChildWithGuardianFactory(
        relationship__guardian__email="enrolled-occurrence-tomorrow-1@example.com",
    )
    child_attended_yesterday_1 = ChildWithGuardianFactory(
        relationship__guardian__email="attended-occurrence-yesterday-1@example.com",
    )
    EnrolmentFactory(occurrence=occurrence_tomorrow_1, child=child_enrolled_tomorrow_1)
    EnrolmentFactory(
        occurrence=occurrence_yesterday_1,
        child=child_attended_yesterday_1,
        attended=True,
    )

    occurrence_tomorrow_2 = OccurrenceFactory(event=event, time=tomorrow)
    occurrence_yesterday_2 = OccurrenceFactory(event=event, time=yesterday)

    child_enrolled_tomorrow_2 = ChildWithGuardianFactory(
        relationship__guardian__email="enrolled-occurrence-tomorrow-2@example.com",
    )
    child_attended_yesterday_2 = ChildWithGuardianFactory(
        relationship__guardian__email="attended-occurrence-yesterday-2@example.com",
    )
    EnrolmentFactory(occurrence=occurrence_tomorrow_2, child=child_enrolled_tomorrow_2)
    EnrolmentFactory(
        occurrence=occurrence_yesterday_2,
        child=child_attended_yesterday_2,
        attended=True,
    )

    child_subscribed_tomorrow_1 = ChildWithGuardianFactory(
        relationship__guardian__email="subscribed-occurrence-tomorrow-1@example.com",
    )
    FreeSpotNotificationSubscriptionFactory(
        child=child_subscribed_tomorrow_1,
        occurrence=occurrence_tomorrow_1,
    )
    child_subscribed_tomorrow_2 = ChildWithGuardianFactory(
        relationship__guardian__email="subscribed-occurrence-tomorrow-2@example.com",
    )
    FreeSpotNotificationSubscriptionFactory(
        child=child_subscribed_tomorrow_2,
        occurrence=occurrence_tomorrow_2,
    )

    another_event = EventFactory(published_at=timezone.now())

    another_event_occurrence_tomorrow = OccurrenceFactory(
        event=another_event, time=tomorrow
    )
    another_event_occurrence_yesterday = OccurrenceFactory(
        event=another_event, time=yesterday
    )

    child_enrolled_another_event = ChildWithGuardianFactory(
        relationship__guardian__email="enrolled-another-event@example.com",
    )
    child_attended_another_event = ChildWithGuardianFactory(
        relationship__guardian__email="attended-another-event@example.com",
    )
    EnrolmentFactory(
        occurrence=another_event_occurrence_tomorrow, child=child_enrolled_another_event
    )
    EnrolmentFactory(
        occurrence=another_event_occurrence_yesterday,
        child=child_attended_another_event,
        attended=True,
    )

    child_subscribed_another_event = ChildWithGuardianFactory(
        relationship__guardian__email="subscribed-another-event@example.com",
    )
    FreeSpotNotificationSubscriptionFactory(
        child=child_subscribed_another_event,
        occurrence=another_event_occurrence_tomorrow,
    )

    child_both_events = ChildWithGuardianFactory(
        relationship__guardian__email="enrolled-both-events@example.com",
    )  # this child should have no invitations
    EnrolmentFactory(
        child=child_both_events,
        occurrence=another_event_occurrence_tomorrow,
    )
    EnrolmentFactory(
        child=child_both_events,
        occurrence=occurrence_tomorrow_1,
    )

    ChildWithGuardianFactory(
        project=ProjectFactory(year=3000),
        relationship__guardian__email="another-project@example.com",
    )

    message = MessageFactory(recipient_selection=recipient_selection)

    if event_selection == "event":
        message.event = event
        message.save()
    elif event_selection == "occurrence_tomorrow_1":
        message.event = event
        message.occurrences.set([occurrence_tomorrow_1])
    elif event_selection == "occurrence_yesterday_1":
        message.event = event
        message.occurrences.set([occurrence_yesterday_1])

    message.send()

    snapshot.assert_match(sorted(m.to for m in mail.outbox))
    assert message.recipient_count == len(mail.outbox)
    assert message.sent_at


@freeze_time("2020-11-11")
@pytest.mark.parametrize("upcoming_occurrence", (True, False))
def test_message_sending_to_invited_group_and_event_groups(upcoming_occurrence):
    """Sending message to invited group when an event group has to be considered.

    Event group consists of two events, with one occurrence each:
        * event 1
            * tomorrow with 1 enrolled child
            * yesterday with 1 attended child
        * event 2
            * yesterday with 1 attended child
    """
    yesterday = timezone.now() - timedelta(days=1)
    tomorrow = timezone.now() + timedelta(days=1)

    event_group = EventGroupFactory(published_at=timezone.now())
    event_1 = EventFactory(event_group=event_group, published_at=timezone.now())
    event_2 = EventFactory(event_group=event_group, published_at=timezone.now())

    occurrence_yesterday_1 = OccurrenceFactory(event=event_1, time=yesterday)
    occurrence_yesterday_2 = OccurrenceFactory(event=event_2, time=yesterday)

    child_attended_yesterday_1 = ChildWithGuardianFactory(
        relationship__guardian__email="attended-occurrence-yesterday-1@example.com",
    )
    child_attended_yesterday_2 = ChildWithGuardianFactory(
        relationship__guardian__email="attended-occurrence-yesterday-2@example.com",
    )

    EnrolmentFactory(
        occurrence=occurrence_yesterday_1,
        child=child_attended_yesterday_1,
        attended=True,
    )
    EnrolmentFactory(
        occurrence=occurrence_yesterday_2,
        child=child_attended_yesterday_2,
        attended=True,
    )

    if upcoming_occurrence:
        occurrence_tomorrow = OccurrenceFactory(event=event_1, time=tomorrow)

        child_enrolled_tomorrow = ChildWithGuardianFactory(
            relationship__guardian__email="enrolled-occurrence-tomorrow@example.com",
        )
        EnrolmentFactory(occurrence=occurrence_tomorrow, child=child_enrolled_tomorrow)

    # Child hasn't attended any events in the event group.
    ChildWithGuardianFactory(
        relationship__guardian__email="should-receive-mail@example.com",
    )

    message = MessageFactory(recipient_selection=Message.INVITED)

    message.send()

    emails = sorted(m.to for m in mail.outbox)

    if upcoming_occurrence:
        assert emails == [["should-receive-mail@example.com"]]
    else:
        assert emails == []
    assert message.recipient_count == len(mail.outbox)
    assert message.sent_at


@freeze_time("2020-11-11")
@pytest.mark.parametrize(
    "enrolled_amount,should_send", [(0, True), (1, True), (2, False), (3, False)]
)
def test_message_sending_to_invited_with_enrolment_limit(
    child_with_user_guardian, enrolled_amount, should_send
):
    """Child who has already enrolled to exactly or more events this year than what is
    allowed by the project's yearly enrolment limit, should not receive an email.
    """
    yesterday = timezone.now() - timedelta(days=1)
    tomorrow = timezone.now() + timedelta(days=1)
    expected_email = child_with_user_guardian.relationships.first().guardian.email

    occurrences = OccurrenceFactory.create_batch(
        enrolled_amount, time=yesterday, event__published_at=yesterday
    )
    for i in range(enrolled_amount):
        EnrolmentFactory(
            child=child_with_user_guardian, occurrence=occurrences[i], attended=True
        )

    # Should receive a message regarding this event (occurrence)
    OccurrenceFactory.create(time=tomorrow, event__published_at=timezone.now())

    message = MessageFactory(recipient_selection=Message.INVITED)

    message.send()

    emails = sorted(m.to for m in mail.outbox)

    if should_send:
        assert emails == [[expected_email]]
    else:
        assert emails == []
    assert message.recipient_count == len(mail.outbox)
    assert message.sent_at
