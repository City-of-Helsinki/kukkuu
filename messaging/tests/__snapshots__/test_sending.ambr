# serializer version: 1
# name: test_cannot_send_message_again_without_force
  list([
    "kukkuu@example.com|['johndoe@example.com']|Otsikko|Ruumisteksti.",
  ])
# ---
# name: test_message_sending[en]
  list([
    "kukkuu@example.com|['johndoe@example.com']|Subject|Body text.",
  ])
# ---
# name: test_message_sending[fi]
  list([
    "kukkuu@example.com|['johndoe@example.com']|Otsikko|Ruumisteksti.",
  ])
# ---
# name: test_message_sending_with_filters[event-all]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-both-events@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[event-attended]
  list([
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[event-enrolled]
  list([
    list([
      'enrolled-both-events@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[event-invited]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[event-subscribed_to_free_spot_notification]
  list([
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[no_event_or_occurrence-all]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-both-events@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[no_event_or_occurrence-attended]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[no_event_or_occurrence-enrolled]
  list([
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-both-events@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[no_event_or_occurrence-invited]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[no_event_or_occurrence-subscribed_to_free_spot_notification]
  list([
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[occurrence_tomorrow_1-all]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-both-events@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[occurrence_tomorrow_1-attended]
  list([
  ])
# ---
# name: test_message_sending_with_filters[occurrence_tomorrow_1-enrolled]
  list([
    list([
      'enrolled-both-events@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[occurrence_tomorrow_1-invited]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[occurrence_tomorrow_1-subscribed_to_free_spot_notification]
  list([
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[occurrence_yesterday_1-all]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-both-events@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[occurrence_yesterday_1-attended]
  list([
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[occurrence_yesterday_1-enrolled]
  list([
  ])
# ---
# name: test_message_sending_with_filters[occurrence_yesterday_1-invited]
  list([
    list([
      'attended-another-event@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-1@example.com',
    ]),
    list([
      'attended-occurrence-yesterday-2@example.com',
    ]),
    list([
      'enrolled-another-event@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'enrolled-occurrence-tomorrow-2@example.com',
    ]),
    list([
      'subscribed-another-event@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-1@example.com',
    ]),
    list([
      'subscribed-occurrence-tomorrow-2@example.com',
    ]),
  ])
# ---
# name: test_message_sending_with_filters[occurrence_yesterday_1-subscribed_to_free_spot_notification]
  list([
  ])
# ---
